### PR TITLE
MSC1802: Remove the '200' value from some federation responses

### LINKED ISSUE #1418: Replace the [200, {}] response for most of the federation requests
See also: https://github.com/matrix-org/synapse/issues/1383

### PR DESCRIPTION
**Pull request by babolivier**
[Rendered](https://github.com/matrix-org/matrix-doc/blob/babolivier/standardised-federation-response-format/proposals/1802-standardised-federation-response-format.md)

Resolves #1418

### CODE REVIEW

**Comment by erikjohnston** [comment]
Cool! Though given we can do this in a backwards compatible way, we should probably do so. Basically, in the same way as #1794 we can fix this in `v2` federation API, and at some point in the future deprecate the `v1` API

**Comment by babolivier** [comment]
> Basically, in the same way as #1794 we can fix this in `v2` federation API, and at some point in the future deprecate the `v1` API

I didn't think about that option. I guess that's indeed better. I'll rewrite my proposal to use it rather than the current proposed solution.

**Comment by babolivier** [comment]
ftr this propsal is ready for review (I hadn't added the in-review label for it as I'd expect someone from the specs core team to do it)

File: proposals/1802-standardised-federation-response-format.md [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+```
+res
+```
+
+Where `res` is the JSON object containing the response to a request directed at
+one of the affected endpoints.
+
+This proposal doesn't address the `PUT
+/_matrix/federation/v1/invite/{roomId}/{eventId}` endpoint since
+[MSC1794](https://github.com/matrix-org/matrix-doc/pull/1794) already takes care
```
Comment (turt2live) [inline proposals/1802-standardised-federation-response-format.md]: Like MSC1794, this should define what the fallback semantics are.

File: proposals/1802-standardised-federation-response-format.md [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+of it.
+
+If a call to any of the `v2` endpoints described in this proposal results in an
+unrecognised request exception, then the sending server should retry the request
```
Comment (turt2live) [inline proposals/1802-standardised-federation-response-format.md]: I really think that this should go into detail for what an unrecognized request is: is it a 404? 400? 500? 302? 501? All of the above?

**Comment by Half-Shot** [comment]
#2102 plans to make a change so that federation endpoints expect canonical JSON rather than JSON be sent along the wire, in the interest of keeping things light for the recipient. I don't want to bring the bikeshed to this proposal, but if we are going to add a new endpoint version then it might be worth merging that rule with this proposal.

**Comment by babolivier** [comment]
Thanks for the heads up @Half-Shot, I'll keep an eye on #2102 to see how it evolves (because it's not clear right now if it's going to end up in the spec).

**Comment by Half-Shot** [comment]
Myself and @anoadragon453 came across the delight of `/send*` endpoints again today /s. Can we get a few more eyes on this?

**Comment by anoadragon453** [comment]
Yes please.

@mscbot fcp merge

**Comment by turt2live** [comment]
I am assuming my past self isn't an idiot, so:

@mscbot reviewed

**Comment by erikjohnston** [comment]
It appears that synapse doesn't return `[200, {...}]` for `/send`, so I think that may be more a spec bug.

**Comment by richvdh** [comment]
yup let's drop /send here and fix it in the spec instead.

Also: can we change the title, which makes the MSC sound much scarier than it is. Just "drop spurious '200' value from federation responses"?

Otherwise lgtm and:

@mscbot reviewed

**Comment by turt2live** [comment]
Have filed https://github.com/matrix-org/matrix-doc/issues/2236

File: proposals/1802-standardised-federation-response-format.md [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+[MSC1794](https://github.com/matrix-org/matrix-doc/pull/1794) already takes care
+of it.
+
+If a call to any of the `v2` endpoints described in this proposal results in an
+unrecognised request exception (i.e. in a response with a 400 or a 404 status
+code), then the sending server should retry the request with the `v1` API.
+
+## Alternative solutions
+
+An alternative solution would be to make the change in the `v1` fedration API,
```
Comment (ara4n) [inline proposals/1802-standardised-federation-response-format.md]:
```suggestion
An alternative solution would be to make the change in the `v1` federation API,
```

**Comment by babolivier** [comment]
> It appears that synapse doesn't return `[200, {...}]` for `/send`, so I think that may be more a spec bug.

@erikjohnston are we sure that other servers (e.g. Construct) don't implement this? If so, I think it mandates a v2 endpoint, otherwise I'm fine with treating it as a spec bug. The latter is more likely (since they possibly wouldn't be able to federate with Synapse in this case?), but we should make sure of that.

**Comment by erikjohnston** [comment]
> > It appears that synapse doesn't return `[200, {...}]` for `/send`, so I think that may be more a spec bug.
> 
> @erikjohnston are we sure that other servers (e.g. Construct) don't implement this? If so, I think it mandates a v2 endpoint, otherwise I'm fine with treating it as a spec bug. The latter is more likely (since they possibly wouldn't be able to federate with Synapse in this case?), but we should make sure of that.

I very much doubt there would be servers in production that implemented federation without testing it against synapse, so I think it should be classified as a spec bug.

**Comment by babolivier** [comment]
Looking at Construct's code (which I think is the only HS implementation that I've seen used in rooms apart from Synapse), I can't see the `[200, {...}]` being used anywhere near `/send` so sgtm.

**Comment by richvdh** [comment]
This was implemented by synapse in https://github.com/matrix-org/synapse/pull/6349, so I think this is just pending a spec PR?

**Comment by turt2live** [comment]
Spec PR: https://github.com/matrix-org/matrix-doc/pull/2547

**Comment by turt2live** [comment]
Merged 🎉

**Pull request merged**